Title feature #19

gdolle opened the pull request:
Pull-request proposal for feature #19 
- Now the title is optional
- Pass the option `title` to display the file name
- Pass a value to this previous option `title:"An example of custom title"` to
  display a custom title.

Bug/Known issue:
- Does not work properly if user write asciidoc pages. See issue #18.
  A workaround for now is to not pass any title, and create it directly in asciidoc.
  For example

``` adoc
.This is the title
[import](path/to/code.ext)
```

Possible improvment:
- Add code caption numbering. (would be useful for pdf format for
  example).

gdolle committed:
Add title feature issue #19

- Now the title is optional
- Pass the option `title` to display the file name
- Pass a value to this previous option `title:"An example of custom title"` to
  display a custom title.

Bug/Known issue:

- Does not work properly if user write asciidoc pages. See issue #18.
A workaround for now is to not pass any title, and create in asciidoc.
For example

```adoc
.This is the title
[import](path/to/code.ext)
```

Possible improvment:

- Add code caption numbering. (would be useful for pdf format for
  example).

azu commented on src/title.js:
```diff
@@ -0,0 +1,43 @@
+ * @return {boolean,string}
+ */
+export function parseTitle(label, fileName) {
+    let title=[false, undefined];
```
Tuple mixed data type is not major in JavaScript.
It will be better that separate `hasTitle` (`regExp.test(string)` is useful) and `parseTitle`?

gdolle replied:
Ok! (with your suggestion, it is not required anymore anyway) ;)

azu commented:
I've written another suggestion to https://github.com/azu/gitbook-plugin-include-codeblock/issues/19#issuecomment-217351006

gdolle committed:
Update parser with key:value object

- See pull request
  https://github.com/azu/gitbook-plugin-include-codeblock/issues/19#issuecomment-217351006
  for more details

gdolle committed:
Fix issue #18 and introduce code counter

- Asciidoc code title fixed #18.
- Introduce a code counter for referencing.

*Example:*

--------------------------
This is a code
[import,title:"my title",id:"ref1"](test.cpp)
and here another code
[import,title:"my title",id:"ref2"](test.cpp)
--------------------------

*Result:*

--------------------------
This is a code

> Code 0: my title
// The code ref1

and here another code

> Code1: my title
// The code ref2
--------------------------

*Cite:*

- asciidoc

"Now I cite both code <<ref1>>, <<ref2>>"

=> "Now I cite both code [Code 0], [Code 1]

NB: Here we have an automatic text replacement.

- markdown

Now I cite both code [Code 0](#ref1), [Code 1](#ref2).

gdolle committed:
Fix parser tests

gdolle committed:
Add comments

gdolle committed:
Update marker tests

gdolle commented:
Code updated!

gdolle committed:
Fix missing markdown number

azu commented on src/parser.js:
```diff
@@ -34,38 +46,113 @@ export function containIncludeCommand(commands = []) {
+        "id":undefined,
+        "marker":undefined
+    };
+    for(var key in keyvals) {
```
Please use `Object.keys(keyvals).forEach(key => { ... })` instead of for..in.
[for...in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) has a trap.

`for...in` is not bad, but prefer use the other.

Related:

[Rule guard-for-in - ESLint - Pluggable JavaScript linter](http://eslint.org/docs/rules/guard-for-in)

azu commented on src/parser.js:
```diff
@@ -34,38 +46,113 @@ export function containIncludeCommand(commands = []) {
+    };
+    for(var key in keyvals) {
+        var keyReg=key;
+        if(key=="marker") {
```
use `===` insteadod `==`.

azu commented on src/parser.js:
```diff
@@ -34,38 +46,113 @@ export function containIncludeCommand(commands = []) {
+    const template = Handlebars.compile(source);
+    // compile with data
+    const output = template(context);
+    // => markdown strings
```
remove this comment. (the `output` is asciidoc or markdown.)

azu commented on src/title.js:
```diff
@@ -0,0 +1,25 @@
+ * @example:
+ *     getTitle(keyValObject)
+ * @param keyValObject
+ * @return {boolean,string}
```
`{boolean|string}` is correct syntax.
- union type: [Use JSDoc: @param](http://usejsdoc.org/tags-param.html)

azu commented on test/marker-test.js:
```diff
@@ -65,82 +65,63 @@ describe("marker", function () {
-                assert(!hasMarker("import test.cpp"));
-                assert(!hasMarker("import, test.cpp"));
-                assert(!hasMarker("import, lang-typescript"));
+                assert(!hasMarker(new Object({title:undefined,id:undefined,marker:undefined})));
```
`new Object` is not needed.
Use only `{ }` objject literal.

``` js
assert(!hasMarker({title:undefined,id:undefined,marker:undefined}));
```

azu commented on test/parser-test.js:
```diff
@@ -1,116 +1,188 @@
             var result = results[0];
-            assert.equal(result.target, "[include, lang-typescript](fixtures/test.ts)");
-            var expected = '> <a name="test.ts" href="fixtures/test.ts">test.ts</a>\n'
+            assert.equal(result.target, "[include, title:\"test.ts\", lang-typescript](fixtures/test.ts)");
```
if you want to avoid writing escaping `\"`, use `'` single quote instead of `"`.

```
'[include, title:"test.ts", lang-typescript](fixtures/test.ts)'
```

is euqal to

```
"[include, title:\"test.ts\", lang-typescript](fixtures/test.ts)"
```

in JavaScript.

azu commented on test/parser-test.js:
```diff
@@ -1,116 +1,188 @@
-            var expected = '> <a name="test.ts" href="fixtures/test.ts">test.ts</a>\n'
+            assert.equal(result.target, "[include, title:\"test.ts\", lang-typescript](fixtures/test.ts)");
+//------------------------------------------------------------------------------
+            var expected =`\
```
❓ 

Is `\` needed?

gdolle replied:
Not really, just a preference to keep a nice indentation.

azu replied:
OK. Thank you for the clarification.

``` js
var output = `\
test
`;
console.log("pre"+output+"post");
/*
pretest
post
*/
```

azu commented on README.md:
```diff
@@ -38,11 +38,37 @@ or
+
+### Title
+
+A title can be added using `title` or `title:<the title>`. In the first case,
```
🔔 

We don't support `title` attribute only pattern.
Need to update README too :)

gdolle replied:
Ah yes forgot to do that :)

azu commented on src/parser.js:
```diff
@@ -34,38 +46,113 @@ export function containIncludeCommand(commands = []) {
+    const source =`\
+{{#if title}}
+    {{#if id}}
+        {% if file.type=="asciidoc" %}
```
🔔 

Aa-ha, `{%` is GitBook template, `{{#if` is handler template.
It is confused for me :)

Please add commet about mixed tempate or separate asciidoc template and markdown template.

gdolle replied:
Ok :)

azu commented on test/title-test.js:
```diff
@@ -0,0 +1,19 @@
+// LICENSE : MIT
+"use strict";
+const assert = require("power-assert");
+//import {getTitle,hasTitle,hasTitleCustom} from "../src/title";
```
Can you remove this comment out?

azu commented on src/parser.js:
```diff
@@ -34,38 +46,113 @@ export function containIncludeCommand(commands = []) {
-    return `> <a name="${fileName}" href="${originalPath}">${fileName}</a>
+/**
+ * generate code from options
+ * @param keyValueObject
```
Please put `{Object}` type to `keyValueObject`.
`@param {Object} keyValueObject` means that `keyValueObject` is object.

azu commented:
@gdolle Thanks for huge update! 👍 
I've reviewed it and I comments: 
- One question ❓ :
  - https://github.com/azu/gitbook-plugin-include-codeblock/pull/20/files#r62426025
- Need to update  🔔 :  
  - https://github.com/azu/gitbook-plugin-include-codeblock/pull/20/files#r62426040
  - https://github.com/azu/gitbook-plugin-include-codeblock/pull/20/files#r62426110
- Other is small things

gdolle committed:
Update second review #20

gdolle commented:
Thanks for your comments. Code updated!

azu commented:
It looks good :)

prudhomm commented:
@azu @gdolle any news on the merging of this PR ?

prudhomm commented:
titles are broken and don't look good. 
Could it be possible to merge this PR and upload a new version of this plugin ?
see here http://book.feelpp.org/content/Tutorial/02-SettingUpEnvironment.html
thx in advance

azu commented:
I wait for updateing https://github.com/azu/gitbook-plugin-include-codeblock/pull/20#discussion_r62426040

> titles are broken and don't look good.

Hmm, Do you use Asciidoc?
https://github.com/feelpp/feelpp-book/blob/master/Tutorial/02-SettingUpEnvironment.adoc

Previously, gitbook-plugin-include-codeblock supported only markdown, but this pull request including support asciidoc.
If it will merge, resolve your issue.

azu commented:
OK. I've merged it.
I will release new version as soon as. 
But it contain breaking change, I will release as 2.0.0-beta #22

prudhomm commented:
@azu yes we use asciidoc. 
should the imported code be colorized (c++) automatically actually ? it does not seem to work but may be I haven't tested/used enough

gdolle commented:
> should the imported code be colorized (c++) automatically actually

@prudhomm See issue #21. Not fixed yet ;)

azu commented:
@prudhomm I've released https://github.com/azu/gitbook-plugin-include-codeblock/releases/tag/1.9.0
Please try it.
( #21 is not fixed yet. )
